virttest/qemu_devices/qbuses.py

```diff
@@ -650,6 +650,118 @@ def _increment_addr(self, addr, last_addr=None):
         return super(QPCIBus, self)._increment_addr(addr, last_addr=last_addr)
 
 
+class QPCIEBus(QPCIBus):
+    """
+    PCIE root bus representation (creates pcie-root-port for virtio pci
+    device by default.)
+    """
+
+    def __init__(self, busid, bus_type, aobject=None,
+                 root_type="pcie-root-port"):
+        """
+        :param busid: id of the bus (pcie.0)
+        :param bus_type: type of the bus (PCIE)
+        :param aobject: Related autotest object (pci.0)
+        :param root_type: root port type
+        """
+        super(QPCIEBus, self).__init__(busid, bus_type, aobject)
+        self.__root_ports = {}
+        self.__root_type = root_type
+
+    def is_direct_plug(self, device):
+        """
+        Justify if the device should be plug to pcie bus directly,
+        the device is not directly plugged only if it is a virtio
+        pci device and the device param direct_plug is "no".
+
+        :param device: the QBaseDevice object
+        :return: bool value for directly plug or not.
+        """
+        pcie_devices = ["virtio-net-pci", "virtio-blk-pci",
+                        "virtio-scsi-pci", "virtio-balloon-pci",
+                        "virtio-serial-pci", "virtio-rng-pci",
+                        "e1000e", "virtio-gpu-device"]
+        if device.get_param("driver") in pcie_devices and (
+                    device.get_param("direct_plug_pci", "no") == "no"):
+            return False
+
+        return True
+
+    def add_root_port(self, addr, root_port=None, multifunction="off"):
```

Comment:
multifunction is not used anywhere yet... do you plan to work on it in the near future? If not it'd probably make sense to just remove that support to avoid false-hope it's supported.

Reply:
@ldoktor I do have test cases for multifunction to be implemented, plan to get the pcie.0 bus and add a root port by directly call its add_root_port() with multifunction="on" in the case itself.